components/core/src/clp/Grep.cpp

```diff
@@ -1063,4 +1011,437 @@ size_t Grep::search(Query const& query, size_t limit, Archive& archive, File& co
+                is_cancel.push_back(false);
+            }
+        }
+    }
+}
+
+void Grep::get_substring_variable_types(
+        uint32_t substr_start,
+        uint32_t substr_end,
+        std::string& schema_search_string,
+        std::vector<bool>& is_greedy_wildcard,
+        std::vector<bool>& is_non_greedy_wildcard,
+        std::vector<bool>& is_cancel,
+        ByteLexer& lexer,
+        bool& contains_wildcard,
+        set<uint32_t>& variable_types
+) {
+    // To determine if a substring could be a variable we convert it to regex,
+    // generate the NFA and DFA for the regex, and intersect the substring DFA with
+    // the compression DFA.
+    std::string regex_search_string;
+    uint32_t pos = 0;
+    for (uint32_t i = substr_start; i <= substr_end; i++) {
+        if (is_cancel[i]) {
+            continue;
+        }
+        auto const& c = schema_search_string[i];
+        if (is_greedy_wildcard[i]) {
+            contains_wildcard = true;
+            regex_search_string += ".*";
+        } else if (is_non_greedy_wildcard[i]) {
+            contains_wildcard = true;
+            regex_search_string += ".";
+        } else if (log_surgeon::SchemaParser::get_special_regex_characters().contains(c)) {
+            regex_search_string += "\\";
+            regex_search_string += c;
+        } else {
+            regex_search_string += c;
+        }
+        pos++;
+    }
+
+    // Generated substring NFA from regex.
+    log_surgeon::Schema substring_schema;
+    // TODO: LogSurgeon should handle resetting this value.
+    log_surgeon::NonTerminal::m_next_children_start = 0;
+    // TODO: could use a forward/reverse lexer in place of intersect a lot of cases.
+    // TODO: NFA creation not optimized at all.
+    substring_schema.add_variable("search", regex_search_string, -1);
+    RegexNFA<RegexNFAByteState> nfa;
+    std::unique_ptr<SchemaAST> schema_ast = substring_schema.release_schema_ast_ptr();
+    for (std::unique_ptr<ParserAST> const& parser_ast : schema_ast->m_schema_vars) {
+        auto* schema_var_ast = dynamic_cast<SchemaVarAST*>(parser_ast.get());
+        ByteLexer::Rule rule(0, std::move(schema_var_ast->m_regex_ptr));
+        rule.add_ast(&nfa);
+    }
+
+    // Generate substring DFA from NFA.
+    // TODO: log-surgeon needs to be refactored to allow direct usage of DFA/NFA.
+    // TODO: DFA creation isn't optimized at all.
+    unique_ptr<RegexDFA<RegexDFAByteState>> dfa2 = lexer.nfa_to_dfa(nfa);
+    unique_ptr<RegexDFA<RegexDFAByteState>> const& dfa1 = lexer.get_dfa();
+
+    // Get variable types in the intersection of substring and compression DFAs.
+    variable_types = dfa1->get_intersect(dfa2);
+}
+
+void Grep::generate_sub_queries(
+        set<QueryLogtype>& query_logtypes,
+        Archive const& archive,
+        ByteLexer& lexer,
+        bool ignore_case,
+        vector<SubQuery>& sub_queries
+) {
+    for (QueryLogtype const& query_logtype : query_logtypes) {
```

Comment:
As discussed offline, let's change this loop to consume each logtype as its processed, and push new logtypes to the end for processing.